Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp

```diff
@@ -284,91 +284,106 @@ bool WebXROpaqueFramebuffer::setupFramebuffer()
-                if (m_attributes.depth)
-                    gl.framebufferRenderbuffer(GL::FRAMEBUFFER, GL::DEPTH_ATTACHMENT, GL::RENDERBUFFER, m_depthStencilBuffer);
-            } else {
-                ASSERT(m_attributes.stencil || m_attributes.depth);
-                gl.framebufferRenderbuffer(GL::FRAMEBUFFER, GL::DEPTH_STENCIL_ATTACHMENT, GL::RENDERBUFFER, m_depthStencilBuffer);
-            }
-        } else {
-            if (m_attributes.stencil) {
-                m_stencilBuffer.ensure(gl);
-                gl.bindRenderbuffer(GL::RENDERBUFFER, m_stencilBuffer);
-                gl.renderbufferStorage(GL::RENDERBUFFER, stencilFormat, m_width, m_height);
-                gl.framebufferRenderbuffer(GL::FRAMEBUFFER, GL::STENCIL_ATTACHMENT, GL::RENDERBUFFER, m_stencilBuffer);
-            }
-            if (m_attributes.depth) {
-                gl.bindRenderbuffer(GL::RENDERBUFFER, m_depthStencilBuffer);
-                gl.renderbufferStorage(GL::RENDERBUFFER, depthFormat, m_width, m_height);
-                gl.framebufferRenderbuffer(GL::FRAMEBUFFER, GL::DEPTH_ATTACHMENT, GL::RENDERBUFFER, m_depthStencilBuffer);
-            }
-        }
+        auto [depthBuffer, stencilBuffer] = allocateDepthStencilStorage(gl, sampleCount, m_width, m_height);
+        bindDepthStencil(gl, depthBuffer, stencilBuffer);
+
+        m_depthStencilBuffer.adopt(gl, depthBuffer);
+        m_stencilBuffer.adopt(gl, stencilBuffer != depthBuffer ? stencilBuffer : 0);
     }
 
     return gl.checkFramebufferStatus(GL::FRAMEBUFFER) == GL::FRAMEBUFFER_COMPLETE;
 }
 
+PlatformGLObject WebXROpaqueFramebuffer::allocateRenderbufferStorage(GraphicsContextGL& gl, GCGLsizei samples, GCGLenum internalFormat, GCGLsizei width, GCGLsizei height)
+{
+    PlatformGLObject renderbuffer = gl.createRenderbuffer();
+    ASSERT(renderbuffer);
+    gl.bindRenderbuffer(GL::RENDERBUFFER, renderbuffer);
+    if (m_context.isWebGL2())
+        gl.renderbufferStorageMultisample(GL::RENDERBUFFER, samples, internalFormat, width, height);
+    else {
+        ASSERT(!samples);
+        gl.renderbufferStorage(GL::RENDERBUFFER, internalFormat, width, height);
+    }
+
+    return renderbuffer;
+}
+
+PlatformGLObject WebXROpaqueFramebuffer::allocateColorStorage(GraphicsContextGL& gl, GCGLsizei samples, GCGLsizei width, GCGLsizei height)
+{
+#if USE(IOSURFACE_FOR_XR_LAYER_DATA) && !PLATFORM(IOS_FAMILY_SIMULATOR)
+    constexpr auto colorFormat = GL::SRGB8_ALPHA8;
+#else
+    constexpr auto colorFormat = GL::RGBA8;
+#endif
+
+    return allocateRenderbufferStorage(gl, samples, colorFormat, width, height);
+}
+
+std::tuple<PlatformGLObject, PlatformGLObject> WebXROpaqueFramebuffer::allocateDepthStencilStorage(GraphicsContextGL& gl, GCGLsizei samples, GCGLsizei width, GCGLsizei height)
+{
+    PlatformGLObject depthBuffer = 0;
+    PlatformGLObject stencilBuffer = 0;
+
+    // FIXME: Does this need to be optional?
```

Comment:
I think there are some non-Apple implementations that don't support the packed format.

Reply:
This code was moved and pre-existed. I've added a `// FIXME:` that we should do something depending upon context or remove the code.